# **Model Training and Feature Engineering**

In [1]:
#### TEMPORARY
import sys
sys.path.append('/home/bac/code/kaggle/kaggle-cmi-detect-behavior/')

In [2]:
import os
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import lightgbm as lgb
import xgboost as xgb
import catboost as cat
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import StratifiedGroupKFold
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import f1_score, balanced_accuracy_score

warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (12, 8)

In [3]:
from src.config import PROJECT_PATH, DATA_PATH, USE_WANDB, WANDB_PROJECT, WANDB_ENTITY
from src.tracking import ExperimentTracker

In [4]:
def display_all(df):
    with pd.option_context("display.max_rows", 100, "display.max_columns", 1000):
        display(df)

In [5]:
# Initialize the experiment tracker
tracker = ExperimentTracker(
    project_path=PROJECT_PATH,
    use_wandb=USE_WANDB,
    wandb_project_name=WANDB_PROJECT,
    wandb_entity=WANDB_ENTITY
)

In [6]:
# Load data
train_sensor = pd.read_csv(os.path.join(os.path.expanduser(DATA_PATH), 'train.csv'))
train_demos = pd.read_csv(os.path.join(os.path.expanduser(DATA_PATH), 'train_demographics.csv'))

# Merge
train_df = pd.merge(train_sensor, train_demos, on='subject', how='left')
display_all(train_df.head())

row_id sequence_type sequence_id  sequence_counter      subject  \
0  SEQ_000007_000000        Target  SEQ_000007                 0  SUBJ_059520   
1  SEQ_000007_000001        Target  SEQ_000007                 1  SUBJ_059520   
2  SEQ_000007_000002        Target  SEQ_000007                 2  SUBJ_059520   
3  SEQ_000007_000003        Target  SEQ_000007                 3  SUBJ_059520   
4  SEQ_000007_000004        Target  SEQ_000007                 4  SUBJ_059520   

                       orientation                                   behavior  \
0  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
1  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
2  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
3  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   
4  Seated Lean Non Dom - FACE DOWN  Relaxes and moves hand to target location   

        phase             gesture     acc_x     acc_y     acc_z     rot_w  \
0  Transition  Cheek - pinch skin  6.683594  6.214844  3.355469  0.134399   
1  Transition  Cheek - pinch skin  6.949219  6.214844  3.125000  0.143494   
2  Transition  Cheek - pinch skin  5.722656  5.410156  5.421875  0.219055   
3  Transition  Cheek - pinch skin  6.601562  3.531250  6.457031  0.297546   
4  Transition  Cheek - pinch skin  5.566406  0.277344  9.632812  0.333557   

      rot_x     rot_y     rot_z      thm_1      thm_2      thm_3      thm_4  \
0 -0.355164 -0.447327 -0.809753  28.943842  31.822186  29.553024  28.592863   
1 -0.340271 -0.428650 -0.824524  29.340816  31.874645  29.791740  28.663383   
2 -0.274231 -0.356934 -0.865662  30.339359  30.935045  30.090014  28.796087   
3 -0.264160 -0.238159 -0.885986  30.543730  27.044001  29.310717  29.018711   
4 -0.218628 -0.063538 -0.914856  29.317265  25.270855  26.808746  29.408604   

       thm_5  tof_1_v0  tof_1_v1  tof_1_v2  tof_1_v3  tof_1_v4  tof_1_v5  \
0  28.310535     131.0     134.0     132.0     135.0      98.0      74.0   
1  28.406172     130.0     138.0     131.0     135.0     101.0      76.0   
2  28.529778     137.0     136.0     147.0     109.0      90.0      81.0   
3  27.402010     143.0     147.0     170.0     127.0     109.0      98.0   
4  27.357603     178.0     191.0     183.0     157.0     146.0     139.0   

   tof_1_v6  tof_1_v7  tof_1_v8  tof_1_v9  tof_1_v10  tof_1_v11  tof_1_v12  \
0      64.0      60.0      -1.0      -1.0      152.0      153.0      141.0   
1      66.0      61.0      -1.0      -1.0      156.0      155.0      141.0   
2      74.0      74.0      -1.0     164.0      165.0      146.0      106.0   
3      95.0      95.0      -1.0     177.0      189.0      177.0      136.0   
4     143.0     148.0      -1.0      -1.0      236.0      238.0      208.0   

   tof_1_v13  tof_1_v14  tof_1_v15  tof_1_v16  tof_1_v17  tof_1_v18  \
0       89.0       68.0       63.0       -1.0       -1.0       -1.0   
1       93.0       74.0       64.0       -1.0       -1.0       -1.0   
2       94.0       77.0       77.0       -1.0       -1.0       -1.0   
3      121.0      107.0      104.0       -1.0       -1.0       -1.0   
4      200.0      185.0      190.0       -1.0       -1.0       -1.0   

   tof_1_v19  tof_1_v20  tof_1_v21  tof_1_v22  tof_1_v23  tof_1_v24  \
0       -1.0      169.0      118.0       86.0       73.0       -1.0   
1       -1.0      165.0      116.0       86.0       75.0      130.0   
2      180.0      140.0      118.0      103.0       92.0       -1.0   
3      202.0      171.0      160.0      141.0      135.0       -1.0   
4      210.0      246.0      225.0      228.0      202.0      149.0   

   tof_1_v25  tof_1_v26  tof_1_v27  tof_1_v28  tof_1_v29  tof_1_v30  \
0       -1.0       -1.0       -1.0       -1.0      147.0      110.0   
1       -1.0       -1.0       -1.0       -1.0      142.0      114.0   
2       -1.0       -1.0       -1.0       -1.0      155.0      119.0   
3       -1.0       -1.0       -1.0    

In [7]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 574945 entries, 0 to 574944
Columns: 348 entries, row_id to elbow_to_wrist_cm
dtypes: float64(334), int64(6), object(8)
memory usage: 1.5+ GB


## Feature Engineering and Experiment Checklist

**Wave 0: Baseline Features (Sequence-Level Aggregates)**
- **Goal**: Establish a rock-solid baseline score.

- Simple statistical aggregations (mean, std, min, max, median, skew) for the core sensor groups (acc_*, rot_*, thm_*) across each entire sequence_id.
- **ToF Handling**: For now, we will simply take the mean of all 320 tof columns as a single feature.

**Wave 1: Phase-Specific Features**

- **Goal**: Capture the distinct behavior within each phase (Transition, Gesture).
- All Wave 0 statistics, but now calculated on a per-phase basis (e.g., mean_acc_x_Transition, mean_acc_x_Gesture). This will significantly expand the feature set.

**Wave 2**: Advanced IMU & Thermopile Features

- **Goal**: Model the physics of the movements.
-   **IMU**: Calculate vector magnitudes (acc_mag, rot_mag), jerk (rate of change of acceleration), and potentially FFT features on the magnitude.  
-   Thermopile: Calculate gradients between adjacent sensors (thm_1-thm_2, etc.) and their statistics per phase.

**Wave 3: Advanced ToF & Demographic Interaction Features**

- **Goal: Extract spatial information from ToF sensors and leverage demographic context.**
- **ToF:** Calculate the percentage of invalid (-1) readings per sensor group. Use PCA or an Autoencoder on the valid ToF readings to create 5-10 latent features.
- **Demographics**: Create interaction features (e.g., mean_acc_mag_Gesture * age, mean_acc_mag_Gesture * height_cm).

**Wave 4: Temporal & Lag Features**

- **Goal**: Capture time-dependent patterns.
- Create rolling window statistics (e.g., 5-step rolling mean) and lag features (e.g., difference between current acc_x and acc_x from 3 steps prior). This requires careful handling of the sequence_counter

## Feature Engineering and Training - Wave 0 - Baseline

In [8]:
def create_baseline_features(df):
    """
    Create Wave 0 features: Simple sequence level aggregations.
    """
    aggs = {
        'acc_x': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'acc_y': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'acc_z': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'rot_w': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'rot_x': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'rot_y': ['mean', 'std', 'min', 'max', 'median', 'skew'],
        'rot_z': ['mean', 'std', 'min', 'max', 'median', 'skew'],
    }
    
    # Add Thermopile aggregations
    for i in range(1, 6):
        aggs[f'thm_{i}'] = ['mean', 'std', 'min', 'max']
        
    # Add time of flight aggregations (simple mean and std for baseline)
    tof_cols = [f'tof_{s}_v{p}' for s in range(1, 6) for p in range(64)]
    # Replace -1 with NaN to correctly calculate stats
    df[tof_cols] = df[tof_cols].replace(-1, np.nan)
    aggs['tof_mean'] = ['mean', 'std']
    df['tof_mean'] = df[tof_cols].mean(axis=1)
    
    # Group by sequence and aggregate and flatten multi index cols
    agg_df = df.groupby('sequence_id').agg(aggs)
    agg_df.columns = ['_'.join(col).strip() for col in agg_df.columns.values]
    
    # Get sequence level metadata (target, subject, etc.)
    meta_df = df.groupby('sequence_id').first()
    
    # Combine aggregated features with metadata
    final_df = pd.concat([meta_df[['subject', 'gesture']], agg_df], axis=1)
    
    # Encode gesture target
    final_df['gesture_encoded'] = final_df['gesture'].astype('category').cat.codes
    
    print(f"Feature engineering complete. Shape of features: {final_df.shape}")
    return final_df

In [9]:
# Create features
features_df = create_baseline_features(train_df)
display_all(features_df.head())

Feature engineering complete. Shape of features: (8151, 67)


subject                   gesture  acc_x_mean  acc_x_std  \
sequence_id                                                                 
SEQ_000007   SUBJ_059520        Cheek - pinch skin    6.153098   1.334155   
SEQ_000008   SUBJ_020948  Forehead - pull hairline    3.400506   1.087142   
SEQ_000013   SUBJ_040282        Cheek - pinch skin   -7.058962   1.295184   
SEQ_000016   SUBJ_052342         Write name on leg    5.524654   1.074108   
SEQ_000018   SUBJ_032165  Forehead - pull hairline    5.363715   1.627637   

             acc_x_min  acc_x_max  acc_x_median  acc_x_skew  acc_y_mean  \
sequence_id                                                               
SEQ_000007    3.613281   9.015625      6.488281   -0.545319    3.915570   
SEQ_000008    1.734375   5.906250      3.437500    0.146452    5.311179   
SEQ_000013   -9.250000  -3.347656     -7.144531    0.518519    2.346182   
SEQ_000016    3.437500   9.378906      5.390625    0.747648   -4.408491   
SEQ_000018    1.964844   6.832031      6.101562   -1.397824    4.109737   

             acc_y_std  acc_y_min  acc_y_max  acc_y_median  acc_y_skew  \
sequence_id                                                              
SEQ_000007    3.048287  -2.019531   6.519531      5.488281   -1.184870   
SEQ_000008    3.268073  -0.222656   8.667969      7.000000   -0.830467   
SEQ_000013    2.564639  -3.273438   4.683594      3.382812   -1.445762   
SEQ_000016    0.598318  -5.718750  -2.960938     -4.492188    0.505319   
SEQ_000018    3.525304  -3.164062   6.718750      6.007812   -1.347944   

             acc_z_mean  acc_z_std  acc_z_min  acc_z_max  acc_z_median  \
sequence_id                                                              
SEQ_000007     5.577782   2.337517   1.093750   9.792969      4.964844   
SEQ_000008     6.581629   2.475402   1.722656  11.074219      5.839844   
SEQ_000013    -6.068544   1.330784 -10.945312  -3.515625     -5.851562   
SEQ_000016    -3.162077   6.139752  -8.078125   8.355469     -6.667969   
SEQ_000018     5.937066   2.104544   4.148438   9.933594      4.761719   

             acc_z_skew  rot_w_mean  rot_w_std  rot_w_min  rot_w_max  \
sequence_id                                                            
SEQ_000007     0.586111    0.263574   0.069033   0.134399   0.379272   
SEQ_000008     0.186346    0.243493   0.064414   0.157593   0.341980   
SEQ_000013    -1.039566    0.392208   0.150629   0.061157   0.540771   
SEQ_000016     0.964846    0.361083   0.041568   0.277527   0.459045   
SEQ_000018     1.040372    0.859159   0.034238   0.828247   0.925049   

             rot_w_median  rot_w_skew  rot_x_mean  rot_x_std  rot_x_min  \
sequence_id                                                               
SEQ_000007       0.254578    0.307637   -0.280817   0.056597  -0.442871   
SEQ_000008       0.226562    0.326730   -0.117145   0.049384  -0.263306   
SEQ_000013       0.439514   -1.359968    0.340804   0.182002   0.140991   
SEQ_000016       0.352234    0.525728   -0.728107   0.207529  -0.893677   
SEQ_000018       0.846283    1.030080    0.177468   0.178091  -0.184204   

             rot_x_max  rot_x_median  rot_x_skew  rot_y_mean  rot_y_std  \
sequence_id                                                               
SEQ_000007   -0.204163     -0.275757   -0.846226   -0.331470   0.175050   
SEQ_000008   -0.050537     -0.097382   -0.937947   -0.342327   0.190164   
SEQ_000013    0.726501      0.258362    1.350803    0.800506   0.090017   
SEQ_000016   -0.384827     -0.857361    0.843064   -0.223281   0.156706   
SEQ_000018    0.305542      0.270111   -1.342766   -0.352176   0.149264   

             rot_y_min  rot_y_max  rot_y_median  rot_y_skew  rot_z_mean  \
sequence_id                                                               
SEQ_000007   -0.478027   0.005066     -0.414978    1.258707   -0.837994   
SEQ_000008   -0.508606  -0.031555     -0.442169    0.830673   -0.875143   
SEQ_000013    0.580505   0.881653      0.838135   -1.593

In [10]:
#Config
EXPERIMENT_NAME = "Baseline-Wave0-LGBM-0"
MODEL_NAME = "LightGBM"
FEATURE_WAVE = "Wave-0"
N_SPLITS = 5
SEED = 42

# Model Params
params = {
    'objective': 'multiclass',
    'num_class': features_df['gesture_encoded'].nunique(),
    'metric': 'multi_logloss',
    'n_estimators': 1000,
    'learning_rate': 0.05,
    'feature_fraction': 0.8,
    'bagging_fraction': 0.8,
    'bagging_freq': 1,
    'lambda_l1': 0.1,
    'lambda_l2': 0.1,
    'num_leaves': 31,
    'verbose': -1,
    'n_jobs': -1,
    'seed': SEED,
    'boosting_type': 'gbdt',
}

# Prep data for CV
X = features_df.drop(columns=['subject', 'gesture', 'gesture_encoded'])
y = features_df['gesture_encoded']
groups = features_df['subject']

In [11]:
# Cross Validation Setup
cv = StratifiedGroupKFold(n_splits=N_SPLITS, shuffle=True, random_state=SEED)

oof_preds = np.zeros((len(features_df), y.nunique()))
oof_true = np.zeros(len(features_df))
fold_scores = []

# Training loop
for fold, (train_idx, val_idx) in enumerate(cv.split(X, y, groups)):
    print(f"--- Fold {fold+1}/{N_SPLITS} ---")
    
    # Split data
    X_train, y_train = X.iloc[train_idx], y.iloc[train_idx]
    X_val, y_val = X.iloc[val_idx], y.iloc[val_idx]
    
    # Init and train model
    model = lgb.LGBMClassifier(**params)
    model.fit(X_train, y_train,
              eval_set=[(X_val, y_val)],
              eval_metric='multi_logloss',
              callbacks=[lgb.early_stopping(100, verbose=False)])
    
    # Predict on validation set
    val_preds = model.predict_proba(X_val)
    oof_preds[val_idx] = val_preds
    oof_true[val_idx] = y_val
    
    # Eval fold performance using balanced_accuracy_score since it is robust to class imbalance
    fold_score = balanced_accuracy_score(y_val, np.argmax(val_preds, axis=1))
    fold_scores.append(fold_score)
    print(f"Fold {fold+1} Balanced Accuracy: {fold_score:.5f}")
    
# Final score and logging
mean_cv_score = np.mean(fold_scores)
print("\n--- CV Summary ---")
print(f"Mean Balanced Accuracy: {mean_cv_score:.5f}")
print(f"Std Dev.: {np.std(fold_scores):.5f}")

# Log the experiment
tracker.log_experiment(
    experiment_name=EXPERIMENT_NAME,
    model_name=MODEL_NAME,
    feature_wave=FEATURE_WAVE,
    cv_score=mean_cv_score,
    params=params,
    notes="Inital baseline model with sequence-level statistical aggregates."
)    

--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.60245
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.52807
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.54332
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.48495
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.52315

--- CV Summary ---
Mean Balanced Accuracy: 0.53639
Std Dev.: 0.03821
Experiment 'Baseline-Wave0-LGBM-0' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


wandb: Currently logged in as: b-a-chaudhry (b-a-chaudhry-) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin
wandb: WARNING Using a boolean value for 'reinit' is deprecated. Use 'return_previous' or 'finish_previous' instead.


Experiment 'Baseline-Wave0-LGBM-0' logged to W&B


#### Analysis:

1. With 18 classes, a random classifier would achieve a balanced accuracy of approximately 1/18 (~5.6%). The current score of ~54% is nearly 10 times better and is a strong confirmation that even the simplest `Wave-0` features contain a strong predictive signal.
2. The standard deviation of 0.0381 across the folds is relatively low, indicating that the model's performance is consistent and not wildly dependent on a specific subset of subjects. 
3. The benchmark score ofo 0.53639 is now the number to beat! 

### Multi-Model Training and CV

Retaining the configurations from `Wave-0`

In [12]:
model_configs = {
    "XGBoost": {
        "model": xgb.XGBClassifier,
        "params": {
            'objective': 'multi:softprob', 'eval_metric': 'mlogloss',
            'n_estimators': 1000, 'learning_rate': 0.05, 'max_depth': 6,
            'subsample': 0.8, 'colsample_bytree': 0.8, 'gamma': 0.1,
            'seed': SEED, 'n_jobs': -1, 'use_label_encoder': False,
        }
    },
    "CatBoost": {
        "model": cat.CatBoostClassifier,
        "params": {
            'iterations': 1000, 'learning_rate': 0.05, 'depth': 6,
            'loss_function': 'MultiClass', 'eval_metric': 'MultiClass',
            'random_seed': SEED, 'verbose': 0,
        }
    },
    "RandomForest": {
        "model": RandomForestClassifier,
        "params": {
            'n_estimators': 200, 'max_depth': 10, 'min_samples_leaf': 5,
            'random_state': SEED, 'n_jobs': -1, 
        }
    }
}

print("="*50)
print("Starting Multi-Model Training on Wave 0 features")
print("="*50)

for model_name, config in model_configs.items():
    EXPERIMENT_NAME = f"Baseline-{FEATURE_WAVE}-{model_name}"
    
    oof_preds = np.zeros((len(features_df), y.nunique()))
    oof_true = np.zeros(len(features_df))
    fold_scores = []
    
    print(f"\n---> Training Model: {model_name} <---")
    cv = StratifiedGroupKFold(n_splits=N_SPLITS, shuffle=True, random_state=SEED)
    for fold, (train_idx, val_idx) in enumerate(cv.split(X, y, groups)):
        print(f"--- Fold {fold+1}/{N_SPLITS} ---")
        
        # Split data
        X_train, y_train = X.iloc[train_idx], y.iloc[train_idx]
        X_val, y_val = X.iloc[val_idx], y.iloc[val_idx]
        
        # Model Initialization and training
        model = config['model'](**config['params'])
        
        # Handle different early stopping mechanisms
        if model_name == 'XGBoost':
            model.fit(X_train, y_train,
                      verbose=False)
        elif model_name == 'CatBoost':
            model.fit(X_train, y_train,
                      eval_set=[(X_val, y_val)],
                      early_stopping_rounds=100,
                      verbose=0)
        else: #Random Forest
            model.fit(X_train, y_train)
        
        # Store preds and ground truth
        val_preds = model.predict_proba(X_val)
        oof_preds[val_idx] = val_preds
        oof_true[val_idx] = y_val
        
        fold_score = balanced_accuracy_score(y_val, np.argmax(val_preds, axis=1))
        fold_scores.append(fold_score)
        print(f"Fold {fold+1} Balanced Accuracy: {fold_score:.5f}")
    
    # Final score and logging
    mean_cv_score = np.mean(fold_scores)
    print(f"\n--- CV Summary for {model_name} ---")
    print(f"Mean Balanced Accuracy: {mean_cv_score:.5f}")
    print(f"Std Dev: {np.std(fold_scores):.5f}\n")
    
    # Log the experiment
    tracker.log_experiment(
        experiment_name=EXPERIMENT_NAME,
        model_name=model_name,
        feature_wave=FEATURE_WAVE,
        cv_score=mean_cv_score,
        params=config['params'],
        notes=f"Multi model baseline run on Wave 0 features using standard CPU training."
    )    

Starting Multi-Model Training on Wave 0 features

---> Training Model: XGBoost <---
--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.62609
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.53573
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.55524
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.50275
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.53145

--- CV Summary for XGBoost ---
Mean Balanced Accuracy: 0.55025
Std Dev: 0.04147

Experiment 'Baseline-Wave-0-XGBoost' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


Experiment 'Baseline-Wave-0-XGBoost' logged to W&B

---> Training Model: CatBoost <---
--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.61826
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.53993
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.55241
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.50637
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.54997

--- CV Summary for CatBoost ---
Mean Balanced Accuracy: 0.55339
Std Dev: 0.03637

Experiment 'Baseline-Wave-0-CatBoost' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


Experiment 'Baseline-Wave-0-CatBoost' logged to W&B

---> Training Model: RandomForest <---
--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.55992
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.46962
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.48397
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.43562
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.47835

--- CV Summary for RandomForest ---
Mean Balanced Accuracy: 0.48549
Std Dev: 0.04082

Experiment 'Baseline-Wave-0-RandomForest' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


Experiment 'Baseline-Wave-0-RandomForest' logged to W&B


#### Analysis

1. Boosting models are out performing the standard Random Forest. Moving forward, we will be focusing on boosting algos.

## Feature Engineering - Wave 1 - Phase Specific Aggregates

`Note` Tree based models can handle NaNs which will naturally appear for sequences missing a phase like 'Pause'.

In [16]:
def create_wave1_features(df):
    """
    Creates Wave-1 features i.e. phase specific aggregates
    """
    imu_cols = ['acc_x', 'acc_y', 'acc_z', 'rot_w', 'rot_x', 'rot_y', 'rot_z']
    thm_cols = [f'thm_{i}' for i in range(1, 6)]
    tof_cols = [f'tof_{s}_v{p}' for s in range(1, 6) for p in range(64)]
    
    # Replace ToF -1 with NaN for correct stats
    df_feat = df.copy()
    df_feat[tof_cols] = df_feat[tof_cols].replace(-1, np.nan)
    
    # Single mean ToF column for all sensors
    df_feat['tof_mean_all_pixels'] = df_feat[tof_cols].mean(axis=1)
    
    # Define aggregations
    aggs = {}
    for col in imu_cols:
        aggs[col] = ['mean', 'std', 'min', 'max', 'skew']
    for col in thm_cols:
        aggs[col] = ['mean', 'std', 'min', 'max']
    aggs['tof_mean_all_pixels'] = ['mean', 'std', 'min', 'max']
    
    # Phase specific aggregations
    phase_agg_df = df_feat.groupby(['sequence_id', 'phase']).agg(aggs)
    
    # Flatten the first level of column names (e.g acc_x --> acc_x_mean)
    phase_agg_df.columns = ['_'.join(col).strip() for col in phase_agg_df.columns.values]
    
    # Unstack the phase level to pivot phases into columns and flatten
    phase_agg_df_unstacked = phase_agg_df.unstack(level='phase')
    phase_agg_df_unstacked.columns = ['_'.join(col).strip() for col in phase_agg_df_unstacked.columns.values] 
    
    # Sequence level meta data (target, subject etc.) to combine with features
    meta_df = df.groupby('sequence_id').first()
    final_df = pd.concat([meta_df[['subject', 'gesture']], phase_agg_df_unstacked], axis=1)
    
    # Encode gesture target
    final_df['gesture_encoded'] = final_df['gesture'].astype('category').cat.codes
    
    print(f"Feature engineering complete. Shape of features: {final_df.shape}")
    return final_df
    

In [17]:
features_df = create_wave1_features(train_df)
display_all(features_df)

Feature engineering complete. Shape of features: (8151, 121)


subject                   gesture  acc_x_mean_Gesture  \
sequence_id                                                              
SEQ_000007   SUBJ_059520        Cheek - pinch skin            6.875854   
SEQ_000008   SUBJ_020948  Forehead - pull hairline            3.449929   
SEQ_000013   SUBJ_040282        Cheek - pinch skin           -7.591276   
SEQ_000016   SUBJ_052342         Write name on leg            5.291590   
SEQ_000018   SUBJ_032165  Forehead - pull hairline            6.235639   
...                  ...                       ...                 ...   
SEQ_065508   SUBJ_027682             Text on phone           -7.286932   
SEQ_065519   SUBJ_050642            Neck - scratch            0.811458   
SEQ_065522   SUBJ_040282     Above ear - pull hair           -9.020573   
SEQ_065526   SUBJ_063447        Cheek - pinch skin            7.894531   
SEQ_065531   SUBJ_039498         Write name on leg            3.813110   

             acc_x_mean_Transition  acc_x_std_Gesture  acc_x_std_Transition  \
sequence_id                                                                   
SEQ_000007                5.227969           0.779424              1.334370   
SEQ_000008                3.353906           0.736728              1.346681   
SEQ_000013               -6.364640           1.143683              1.162073   
SEQ_000016                5.818142           0.940102              1.174814   
SEQ_000018                3.881445           0.241211              1.901476   
...                            ...                ...                   ...   
SEQ_065508                3.387445           0.393596              7.357197   
SEQ_065519                3.822917           0.862646              3.676374   
SEQ_065522                1.589518           0.501440              5.849199   
SEQ_065526                5.653646           0.252330              2.180823   
SEQ_065531                5.267299           0.522013              1.316939   

             acc_x_min_Gesture  acc_x_min_Transition  acc_x_max_Gesture  \
sequence_id                                                               
SEQ_000007            5.492188              3.613281           9.015625   
SEQ_000008            1.734375              1.886719           4.949219   
SEQ_000013           -9.250000             -7.714844          -5.226562   
SEQ_000016            3.859375              3.437500           9.378906   
SEQ_000018            5.718750              1.964844           6.832031   
...                        ...                   ...                ...   
SEQ_065508           -8.058594             -9.359375          -6.410156   
SEQ_065519           -0.628906             -1.433594           1.937500   
SEQ_065522          -10.023438             -9.339844          -7.847656   
SEQ_065526            7.324219              2.414062           8.390625   
SEQ_065531            2.546875              3.351562           4.652344   

             acc_x_max_Transition  acc_x_skew_Gesture  acc_x_skew_Transition  \
sequence_id                                                                    
SEQ_000007               7.292969            0.671795               0.119323   
SEQ_000008               5.906250           -0.292417               0.278393   
SEQ_000013              -3.347656            0.395099               1.138648   
SEQ_000016               7.925781            2.323306              -0.451960   
SEQ_000018               6.332031            0.194196               0.300721   
...                           ...                 ...                    ...   
SEQ_065508              12.167969            0.251036              -0.565475   
SEQ_065519               9.632812           -0.364536              -0.477535   
SEQ_065522               7.437500            0.315421              -1.283318   
SEQ_065526               9.343750           -0.491839               0.135434   
SEQ_065531               8.566406           -0.468476               0.868013   

         

In [18]:
FEATURE_WAVE = "Wave 1"
N_SPLITS = 5
SEED = 42

In [19]:
# --- Prepare data for CV ---
X = features_df.drop(columns=['subject', 'gesture', 'gesture_encoded'])
y = features_df['gesture_encoded']
groups = features_df['subject']

In [22]:
# --- Updated Model Configurations (Top 3) ---
model_configs = {
    "LightGBM": {
        "model": lgb.LGBMClassifier,
        "params": { 'objective': 'multiclass', 'num_class': features_df['gesture_encoded'].nunique(), 'metric': 'multi_logloss', 'n_estimators': 1000, 'learning_rate': 0.05, 'feature_fraction': 0.8, 'bagging_fraction': 0.8, 'bagging_freq': 1, 'lambda_l1': 0.1, 'lambda_l2': 0.1, 'num_leaves': 31, 'verbose': -1, 'n_jobs': -1, 'seed': SEED, 'boosting_type': 'gbdt' }
    },
    "XGBoost": {
        "model": xgb.XGBClassifier,
        "params": { 'objective': 'multi:softprob', 'eval_metric': 'mlogloss', 'n_estimators': 1000, 'learning_rate': 0.05, 'max_depth': 6, 'subsample': 0.8, 'colsample_bytree': 0.8, 'gamma': 0.1, 'seed': SEED, 'n_jobs': -1, 'use_label_encoder': False }
    },
    "CatBoost": {
        "model": cat.CatBoostClassifier,
        "params": { 'iterations': 1000, 'learning_rate': 0.05, 'depth': 6, 'loss_function': 'MultiClass', 'eval_metric': 'MultiClass', 'random_seed': SEED, 'verbose': 0 }
    }
}

In [23]:
# --- Cross-Validation Loop for Each Model ---
print("="*50)
print(f"Starting Multi-Model CPU-Based Training on {FEATURE_WAVE} Features")
print("="*50)

for model_name, config in model_configs.items():
    
    EXPERIMENT_NAME = f"{FEATURE_WAVE}-{model_name}-CPU"
    fold_scores = []

    print(f"\n---> Training Model: {model_name} <---")
    
    cv = StratifiedGroupKFold(n_splits=N_SPLITS, shuffle=True, random_state=SEED)
    for fold, (train_idx, val_idx) in enumerate(cv.split(X, y, groups)):
        print(f"--- Fold {fold+1}/{N_SPLITS} ---")
        
        X_train, y_train = X.iloc[train_idx], y.iloc[train_idx]
        X_val, y_val = X.iloc[val_idx], y.iloc[val_idx]
        
        model = config['model'](**config['params'])
        
        if model_name == 'LightGBM':
            model.fit(X_train, y_train, eval_set=[(X_val, y_val)], callbacks=[lgb.early_stopping(100, verbose=False)])
        elif model_name == 'CatBoost':
            model.fit(X_train, y_train, eval_set=[(X_val, y_val)], early_stopping_rounds=100)
        elif model_name == 'XGBoost':
            model.fit(X_train, y_train, verbose=False)
            
        val_preds = model.predict_proba(X_val)
        fold_score = balanced_accuracy_score(y_val, np.argmax(val_preds, axis=1))
        fold_scores.append(fold_score)
        print(f"Fold {fold+1} Balanced Accuracy: {fold_score:.5f}")

    mean_cv_score = np.mean(fold_scores)
    print(f"\n--- CV Summary for {model_name} ---")
    print(f"Mean Balanced Accuracy: {mean_cv_score:.5f}")
    print(f"Std Dev: {np.std(fold_scores):.5f}\n")
    
    tracker.log_experiment(
        experiment_name=EXPERIMENT_NAME,
        model_name=model_name,
        feature_wave=FEATURE_WAVE,
        cv_score=mean_cv_score,
        params=config['params'],
        notes="Phase-specific statistical aggregates."
    )

Starting Multi-Model CPU-Based Training on Wave 1 Features

---> Training Model: LightGBM <---
--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.68979
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.63108
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.65937
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.58927
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.59668

--- CV Summary for LightGBM ---
Mean Balanced Accuracy: 0.63324
Std Dev: 0.03783

Experiment 'Wave 1-LightGBM-CPU' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


Experiment 'Wave 1-LightGBM-CPU' logged to W&B

---> Training Model: XGBoost <---
--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.70552
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.62616
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.65261
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.60229
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.61356

--- CV Summary for XGBoost ---
Mean Balanced Accuracy: 0.64003
Std Dev: 0.03678

Experiment 'Wave 1-XGBoost-CPU' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


Experiment 'Wave 1-XGBoost-CPU' logged to W&B

---> Training Model: CatBoost <---
--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.71385
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.65075
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.64162
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.60518
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.61343

--- CV Summary for CatBoost ---
Mean Balanced Accuracy: 0.64497
Std Dev: 0.03839

Experiment 'Wave 1-CatBoost-CPU' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


Experiment 'Wave 1-CatBoost-CPU' logged to W&B


#### Analysis

1. Massive Performance Improvement: This was not an incremental gain; it was a huge leap across the board. The phase-specific features provided a much richer, more granular signal for the models to learn from.

2. Hypothesis Confirmed: The core hypothesis for Wave 1 was that the statistical properties of the sensors are fundamentally different and more informative during the Transition and Gesture phases. A ~9.2% absolute improvement in our best model's accuracy validates this.

3. **CatBoost Solidifies its Lead**: CatBoost continues to be the top-performing algorithm on this dataset, achieving a new best score of **0.64497**. This will be our new benchmark to beat. The stability remains excellent, with a standard deviation of ~0.038, indicating the performance is consistent across all folds.

## Feature Engineering - Wave 2 - Physically Relevant Features From IMU and Thermopile Sensors. 

IMU: Vector magnitudes (acc_mag, rot_mag) and jerk (the derivative of acceleration magnitude).

Thermopile: Temperature gradients between adjacent sensors.

In [24]:
def create_wave2_features(df):
    """ 
    Creates wave 2 features: Adds advanced IMU (magnitude, jerk) and Thermopile (gradients)
    features before performing phase specifc aggregation.
    """
    df_feat = df.copy()
    
    # Create sensor derived features
    # IMU magnitudes
    df_feat['acc_mag'] = np.sqrt(df_feat['acc_x']**2 + df_feat['acc_y']**2 + df_feat['acc_z']**2)
    df_feat['rot_mag'] = np.sqrt(df_feat['rot_w']**2 + df_feat['rot_x']**2 + df_feat['rot_y']**2 + df_feat['rot_z']**2)
    
    # IMU Jerk (Rate of change of acceleration magnitude)
    # Group by sequence_id to prevent calculating diff across different sequences
    df_feat['jerk'] = df_feat.groupby('sequence_id')['acc_mag'].diff().fillna(0)
    
    # Thermopile gradients
    for i in range(1, 5):
        df_feat[f'thm_grad_{i}_{i+1}'] = df_feat[f'thm_{i}'] - df_feat[f'thm_{i+1}']
    
    # Repeat Wave 1 Phase-Specific Aggregations
    # Define columns to aggregate
    imu_derived_cols = ['acc_mag', 'rot_mag', 'jerk']
    thm_grad_cols = [f'thm_grad_{i}_{i+1}' for i in range(1, 5)]
    original_sensor_cols = [c for c in df.columns if 'acc_' in c or 'rot_' in c or 'thm_' in c]
    tof_cols = [f'tof_{s}_v{p}' for s in range(1, 6) for p in range(64)]

    df_feat[tof_cols] = df_feat[tof_cols].replace(-1, np.nan)
    df_feat['tof_mean_all_pixels'] = df_feat[tof_cols].mean(axis=1)

    aggs = {}
    for col in original_sensor_cols + imu_derived_cols + thm_grad_cols:
        aggs[col] = ['mean', 'std', 'min', 'max', 'skew']
    aggs['tof_mean_all_pixels'] = ['mean', 'std', 'min', 'max']

    phase_agg_df = df_feat.groupby(['sequence_id', 'phase']).agg(aggs)
    phase_agg_df.columns = ['_'.join(col).strip() for col in phase_agg_df.columns.values]
    phase_agg_df_unstacked = phase_agg_df.unstack(level='phase')
    phase_agg_df_unstacked.columns = ['_'.join(col).strip() for col in phase_agg_df_unstacked.columns.values]
    
    meta_df = df.groupby('sequence_id').first()
    final_df = pd.concat([meta_df[['subject', 'gesture']], phase_agg_df_unstacked], axis=1)
    final_df['gesture_encoded'] = final_df['gesture'].astype('category').cat.codes
    
    print(f"Feature engineering complete. Shape of features: {final_df.shape}")
    return final_df

In [26]:
features_df = create_wave2_features(train_df)
display_all(features_df.head())

Feature engineering complete. Shape of features: (8151, 201)


subject                   gesture  acc_x_mean_Gesture  \
sequence_id                                                              
SEQ_000007   SUBJ_059520        Cheek - pinch skin            6.875854   
SEQ_000008   SUBJ_020948  Forehead - pull hairline            3.449929   
SEQ_000013   SUBJ_040282        Cheek - pinch skin           -7.591276   
SEQ_000016   SUBJ_052342         Write name on leg            5.291590   
SEQ_000018   SUBJ_032165  Forehead - pull hairline            6.235639   

             acc_x_mean_Transition  acc_x_std_Gesture  acc_x_std_Transition  \
sequence_id                                                                   
SEQ_000007                5.227969           0.779424              1.334370   
SEQ_000008                3.353906           0.736728              1.346681   
SEQ_000013               -6.364640           1.143683              1.162073   
SEQ_000016                5.818142           0.940102              1.174814   
SEQ_000018                3.881445           0.241211              1.901476   

             acc_x_min_Gesture  acc_x_min_Transition  acc_x_max_Gesture  \
sequence_id                                                               
SEQ_000007            5.492188              3.613281           9.015625   
SEQ_000008            1.734375              1.886719           4.949219   
SEQ_000013           -9.250000             -7.714844          -5.226562   
SEQ_000016            3.859375              3.437500           9.378906   
SEQ_000018            5.718750              1.964844           6.832031   

             acc_x_max_Transition  acc_x_skew_Gesture  acc_x_skew_Transition  \
sequence_id                                                                    
SEQ_000007               7.292969            0.671795               0.119323   
SEQ_000008               5.906250           -0.292417               0.278393   
SEQ_000013              -3.347656            0.395099               1.138648   
SEQ_000016               7.925781            2.323306              -0.451960   
SEQ_000018               6.332031            0.194196               0.300721   

             acc_y_mean_Gesture  acc_y_mean_Transition  acc_y_std_Gesture  \
sequence_id                                                                 
SEQ_000007             5.640137               1.708125           0.582177   
SEQ_000008             7.261364               3.472433           0.824113   
SEQ_000013             3.751562               0.513077           0.571292   
SEQ_000016            -4.744830              -3.984954           0.441674   
SEQ_000018             6.068704               0.779492           0.211266   

             acc_y_std_Transition  acc_y_min_Gesture  acc_y_min_Transition  \
sequence_id                                                                  
SEQ_000007               3.493182           4.683594             -2.019531   
SEQ_000008               3.638059           4.718750             -0.222656   
SEQ_000013               2.984959           2.656250             -3.273438   
SEQ_000016               0.494274          -5.718750             -4.875000   
SEQ_000018               4.005993           5.566406             -3.164062   

             acc_y_max_Gesture  acc_y_max_Transition  acc_y_skew_Gesture  \
sequence_id                                                                
SEQ_000007            6.519531              6.214844           -0.177613   
SEQ_000008            8.667969              7.976562           -0.761132   
SEQ_000013            4.683594              3.957031            0.011108   
SEQ_000016           -3.113281             -2.960938            1.184362   
SEQ_000018            6.449219              6.718750           -0.492244   

             acc_y_skew_Transition  acc_z_mean_Gesture  acc_z_mean_Transition  \
sequence_id                                                                     
SEQ_000007                0.188431            4.259521               7.2651

In [27]:
FEATURE_WAVE = "Wave 2"
N_SPLITS = 5
SEED = 42

# --- Prepare data for CV ---
X = features_df.drop(columns=['subject', 'gesture', 'gesture_encoded'])
y = features_df['gesture_encoded']
groups = features_df['subject']

In [28]:
# --- Cross-Validation Loop ---
print("="*50)
print(f"Starting Multi-Model CPU-Based Training on {FEATURE_WAVE} Features")
print("="*50)

for model_name, config in model_configs.items():
    EXPERIMENT_NAME = f"{FEATURE_WAVE}-{model_name}-CPU"
    fold_scores = []
    
    print(f"\n---> Training Model: {model_name} <---")
    cv = StratifiedGroupKFold(n_splits=N_SPLITS, shuffle=True, random_state=SEED)
    
    for fold, (train_idx, val_idx) in enumerate(cv.split(X, y, groups)):
        print(f"--- Fold {fold+1}/{N_SPLITS} ---")
        X_train, y_train = X.iloc[train_idx], y.iloc[train_idx]
        X_val, y_val = X.iloc[val_idx], y.iloc[val_idx]
        model = config['model'](**config['params'])
        if model_name == 'LightGBM':
            model.fit(X_train, y_train, eval_set=[(X_val, y_val)], callbacks=[lgb.early_stopping(100, verbose=False)])
        elif model_name == 'CatBoost':
            model.fit(X_train, y_train, eval_set=[(X_val, y_val)], early_stopping_rounds=100, verbose=0)
        elif model_name == 'XGBoost':
            model.fit(X_train, y_train, verbose=False)
        val_preds = model.predict_proba(X_val)
        fold_score = balanced_accuracy_score(y_val, np.argmax(val_preds, axis=1))
        fold_scores.append(fold_score)
        print(f"Fold {fold+1} Balanced Accuracy: {fold_score:.5f}")
    
    mean_cv_score = np.mean(fold_scores)
    print(f"\n--- CV Summary for {model_name} ---")
    print(f"Mean Balanced Accuracy: {mean_cv_score:.5f}")
    print(f"Std Dev: {np.std(fold_scores):.5f}\n")
    
    tracker.log_experiment(
        experiment_name=EXPERIMENT_NAME, model_name=model_name, feature_wave=FEATURE_WAVE,
        cv_score=mean_cv_score, params=config['params'],
        notes="Added IMU (mag, jerk) and Thermopile (gradient) features before phase aggregation."
    )

Starting Multi-Model CPU-Based Training on Wave 2 Features

---> Training Model: LightGBM <---
--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.69788
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.63267
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.66461
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.59172
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.60961

--- CV Summary for LightGBM ---
Mean Balanced Accuracy: 0.63930
Std Dev: 0.03811

Experiment 'Wave 2-LightGBM-CPU' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


Experiment 'Wave 2-LightGBM-CPU' logged to W&B

---> Training Model: XGBoost <---
--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.70817
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.64598
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.66839
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.61762
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.62309

--- CV Summary for XGBoost ---
Mean Balanced Accuracy: 0.65265
Std Dev: 0.03309

Experiment 'Wave 2-XGBoost-CPU' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


Experiment 'Wave 2-XGBoost-CPU' logged to W&B

---> Training Model: CatBoost <---
--- Fold 1/5 ---
Fold 1 Balanced Accuracy: 0.73970
--- Fold 2/5 ---
Fold 2 Balanced Accuracy: 0.65958
--- Fold 3/5 ---
Fold 3 Balanced Accuracy: 0.65665
--- Fold 4/5 ---
Fold 4 Balanced Accuracy: 0.62008
--- Fold 5/5 ---
Fold 5 Balanced Accuracy: 0.61642

--- CV Summary for CatBoost ---
Mean Balanced Accuracy: 0.65849
Std Dev: 0.04437

Experiment 'Wave 2-CatBoost-CPU' logged to ~/code/kaggle/kaggle-cmi-detect-behavior/experiment_log.csv


Experiment 'Wave 2-CatBoost-CPU' logged to W&B
